Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/PolDiffILLReduction.py

```diff
@@ -0,0 +1,589 @@
+
+        for entry_no in range(mtd[ws].getNumberOfEntries()):
+            if entry_no % 2 != 0:
+                continue
+            polarisation_entry_no = int(entry_no/2)
+            if singleCorrectionPerPOL:
+                polarisation_entry_no = int(polarisation_entry_no % nPolarisations)
+            phi = mtd[pol_eff_ws][polarisation_entry_no].name()
+            intensity_0 = mtd[ws][entry_no].name()
+            intensity_1 = mtd[ws][entry_no+1].name()
+            tmp_names = [intensity_0 + '_tmp', intensity_1 + '_tmp']
+
+            nominator = ((1.0-mtd[phi])*(1.0-fp) + fp*(1+mtd[phi])) * mtd[intensity_0] \
+                - (1.0-mtd[phi]) * mtd[intensity_1]
+            denominator = 2.0 * fp * mtd[phi]
+            Divide(LHSWorkspace=nominator, RHSWorkspace=denominator,
+                   OutputWorkspace=tmp_names[0])
+            nominator = (1+mtd[phi])*mtd[intensity_1] \
+                - ( (1+mtd[phi])*(1-fp) - fp*(1-mtd[phi]) )*mtd[intensity_0]
+            denominator = 2.0 * fp * mtd[phi]
+            Divide(LHSWorkspace=nominator, RHSWorkspace=denominator,
+                   OutputWorkspace=tmp_names[1])
+
+            RenameWorkspace(tmp_names[0], intensity_0)
+            RenameWorkspace(tmp_names[1], intensity_1)
+
+        DeleteWorkspaces(WorkspaceList=[nominator, denominator])
+        return ws
+
+    def _read_experiment_properties(self, ws):
+        """Reads the user-provided dictionary that contains sample geometry (type, dimensions) and experimental conditions,
+         such as the beam size and calculates derived parameters."""
+        self._sampleAndEnvironmentProperties = self.getProperty('SampleAndEnvironmentProperties').value
+        if 'initial_energy' not in self._sampleAndEnvironmentProperties:
+            h = physical_constants['Planck constant'][0]  # in m^2 kg^2 / s^2
+            neutron_mass = physical_constants['neutron mass'][0]  # in kg
+            wavelength = mtd[ws][0].getRun().getLogData('monochromator.wavelength').value * 1e-10  # in m
+            joules_to_mev = 1e3 / physical_constants['electron volt'][0]
+            self._sampleAndEnvironmentProperties['InitialEnergy'] = joules_to_mev * math.pow(h / wavelength, 2) / (2 * neutron_mass)
+
+    def _enforce_uniform_units(self, origin_ws, target_ws):
+        for entry_tuple in zip(mtd[origin_ws], mtd[target_ws]):
+            entry_origin, entry_target = entry_tuple
+            if entry_origin.YUnit() != entry_target.YUnit():
+                entry_target.setYUnit(entry_origin.YUnit())
+
+    def _apply_self_attenuation_correction(self, sample_ws, container_ws):
+        """Applies the self-attenuation correction based on the Palmaan-Pings Monte-Carlo calculation, taking into account
+        the sample's material, shape, and dimensions."""
+        geometry_type = self.getPropertyValue('SampleGeometry')
+
+        kwargs = {}
```

Comment:
Just to check, is there still some key renaming due?

Reply:
I've had a second look and added two unit tests for workflows with PPMC that lead to some small modifications of names used. Now everything is conformant with PPMC and sets the correct units (specifically the Number density in Formula Units).